### PR TITLE
Generalize exponential

### PR DESCRIPTION
**Pull request by t4c1**
## Summary
Generalizes exponential distribution related functions to accept Eigen Expressions.

## Tests
Tested using expression testing framework.

## Side Effects
None.

## Release notes
Generalized exponential distribution related functions to accept Eigen Expressions.

## Checklist

- [x] Math issue #1470 

- [x] Copyright holder: Tadej Ciglarič
    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/prim/prob/exponential_lpdf.hpp [hunk 1]
```diff
@@ -46,51 +47,62 @@ template <bool propto, typename T_y, typename T_inv_scale>
...
+    using beta_val_array = Eigen::Array<beta_val_scalar, Eigen::Dynamic, 1>;
+    if (is_vector<T_y>::value && !is_vector<T_inv_scale>::value) {
+      ops_partials.edge1_.partials_ = T_partials_array::Constant(
+          size(y), -static_cast<T_partials_return>(
+                       forward_as<beta_val_scalar>(beta_val)));
+    } else if (is_vector<T_inv_scale>::value) {
+      ops_partials.edge1_.partials_ = -forward_as<beta_val_array>(beta_val)
+                                           .template cast<T_partials_return>();
+    } else {
+      forward_as<internal::broadcast_array<T_partials_return>>(
```
Comment (bbbales2) [inline stan/math/prim/prob/exponential_lpdf.hpp:97]: Hmm does this need to be this complicated? Surely there is a simpler way. Maybe declaring beta_val_scalar a different type?

File: stan/math/prim/prob/exponential_lpdf.hpp [hunk 2]
```diff
@@ -46,51 +47,62 @@ template <bool propto, typename T_y, typename T_inv_scale>
...
-      ops_partials.edge2_.partials_[n] += inv(beta_dbl) - y_dbl;
+  if (!is_constant_all<T_y>::value) {
+    using beta_val_scalar = scalar_type_t<decltype(beta_val)>;
+    using beta_val_array = Eigen::Array<beta_val_scalar, Eigen::Dynamic, 1>;
+    if (is_vector<T_y>::value && !is_vector<T_inv_scale>::value) {
+      ops_partials.edge1_.partials_ = T_partials_array::Constant(
+          size(y), -static_cast<T_partials_return>(
+                       forward_as<beta_val_scalar>(beta_val)));
+    } else if (is_vector<T_inv_scale>::value) {
+      ops_partials.edge1_.partials_ = -forward_as<beta_val_array>(beta_val)
```
Comment (bbbales2) [inline stan/math/prim/prob/exponential_lpdf.hpp:94]: Yes I don't like these casts and such. Also I am reviewing this bottom to top if these comments seem out of order lol.

**Review by bbbales2 — CHANGES_REQUESTED** [review]
I don't like all the casting near the forwarding. Is it necessary?

File: stan/math/prim/prob/exponential_lpdf.hpp (hunk 1 above)
Comment (t4c1) [reply stan/math/prim/prob/exponential_lpdf.hpp:95]: Hmm, maybe not? Let me test removing those.

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Ack I did a review yesterday and never hit submit sorry.

File: stan/math/prim/prob/exponential_lccdf.hpp [hunk 3]
```diff
@@ -15,32 +16,60 @@ template <typename T_y, typename T_inv_scale>
...
+  if (!is_constant_all<T_y>::value) {
+    using beta_val_scalar = scalar_type_t<decltype(beta_val)>;
+    using beta_val_array = Eigen::Array<beta_val_scalar, Eigen::Dynamic, 1>;
+    if (is_vector<T_y>::value && !is_vector<T_inv_scale>::value) {
+      ops_partials.edge1_.partials_ = T_partials_array::Constant(
+          size(y), -forward_as<beta_val_scalar>(beta_val));
+    } else if (is_vector<T_inv_scale>::value) {
+      ops_partials.edge1_.partials_ = -forward_as<beta_val_array>(beta_val);
+    } else {
+      forward_as<internal::broadcast_array<T_partials_return>>(
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/prim/prob/exponential_lccdf.hpp:56]: Unfortunately we do need those, as Eigen matrices do not have an assignment operator for assigning a scalar (and partials_ can be a `Eigen::VectorXd`, even if not in this else).

[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/prim/prob/exponential_lccdf.hpp:56]: Can we just do a call tomorrow and talk through this? I'm having trouble figuring it out.

[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/prim/prob/exponential_lccdf.hpp:56]: 7AM Pacific?

[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/prim/prob/exponential_lccdf.hpp:56]: I can probably make it.

**Comment by bbbales2** [comment]
Merge conflicts appeared but I approve of this whenever those are resolved and tests pass

**Review by bbbales2 — APPROVED** [review]

**Pull request merged**